src/services/download/mod.rs

```diff
@@ -0,0 +1,171 @@
+//! Service which handles all downloading from multiple kinds of sources.
+//!
+//! The sources are described on
+//! [https://docs.sentry.io/workflow/debug-files/#symbol-servers](https://docs.sentry.io/workflow/debug-files/#symbol-servers)
+
+use std::io::Write;
+use std::path::PathBuf;
+use std::string::ToString;
+
+use failure::ResultExt;
+use futures::compat::Future01CompatExt;
+use futures::future::{FutureExt, TryFutureExt};
+use futures01::future;
+use futures01::prelude::*;
+
+use crate::utils::futures::RemoteThread;
+
+mod filesystem;
+mod gcs;
+mod http;
+mod s3;
+mod sentry;
+mod types;
+
+pub use self::types::{DownloadError, DownloadErrorKind, DownloadStatus, DownloadStream};
+pub use crate::sources::{SentryFileId, SourceFileId, SourceLocation};
+
+/// A service which can download files from a [`SourceConfig`].
+///
+/// The service is rather simple on the outside but will one day control
+/// rate limits and the concurrency it uses.
+///
+/// [`SourceConfig`]: ../../types/enum.SourceConfig.html
+#[derive(Debug, Clone)]
+pub struct Downloader {
+    worker: RemoteThread,
+}
+
+impl Downloader {
+    pub fn new(worker: RemoteThread) -> Self {
+        Self { worker }
+    }
+
+    /// Download a file from a source and store it on the local filesystem.
+    ///
+    /// This does not do any deduplication of requests, every requested file is
+    /// freshly downloaded.
+    ///
+    /// # Arguments
+    ///
+    /// `source` - The source to download from.
+    ///
+    /// `dest` - Pathname of filename to save the downloaded file into.  The
+    ///    file will be created if it does not exist and truncated if it does.
+    ///    On successful completion the file's contents will be the download
+    ///    result.  In case of any error the file's contents is considered
+    ///    garbage.
+    ///
+    /// # Return value
+    ///
+    /// On success returns `Some(dest)`, if the download failed e.g. due to an
+    /// HTTP 404, `None` is returned.  If there is an error during the
+    /// downloading [`DownloadError`] is returned.
+    ///
+    /// [`DownloadError`]: types/struct.DownloadError.html
+    pub fn download(
+        &self,
+        source: SourceFileId,
+        dest: PathBuf,
+    ) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError> + Send + 'static> {
+        let fut03 = self.worker.spawn(|| async move {
```

Comment:
This toplevel spawn should probably be recording metrics for the "futures.wait_time" metric.